From 4893b6c9dde3b9fafffbdca5ca97b14b79e9634a Mon Sep 17 00:00:00 2001
From: Carlos Munoz <cmunoz@caviumnetworks.com>
Date: Thu, 3 Sep 2015 16:45:37 -0700
Subject: [PATCH 146/184] netdev: octeon3-ethernet: Linearize skbs with more
 than 11 segments.

Source: Cavium Networks, Inc.
MR: 00000
Type: Integration
Disposition: Merged from Octeon Tree
ChangeID: aae39dc3586f9de4a63bee5d2b1d88e9997b2fd4
Description:

Signed-off-by: Carlos Munoz <cmunoz@caviumnetworks.com>
Signed-off-by: Corey Minyard <cminyard@mvista.com>
[Original patch taken from octeon-linux-kernel-patches-SDK-3.1.2-release]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/net/ethernet/octeon/octeon3-ethernet.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/octeon/octeon3-ethernet.c b/drivers/net/ethernet/octeon/octeon3-ethernet.c
index de90372..8335853 100644
--- a/drivers/net/ethernet/octeon/octeon3-ethernet.c
+++ b/drivers/net/ethernet/octeon/octeon3-ethernet.c
@@ -1928,6 +1928,7 @@ static int octeon3_eth_ndo_start_xmit(struct sk_buff *skb, struct net_device *ne
 	    skb_shared(skb) == 0 &&
 	    skb_cloned(skb) == 0 &&
 	    frag_count == 0 &&
+	    skb_shinfo(skb)->nr_frags == 0 &&
 	    skb_headroom(skb) >= 128 &&
 	    skb_end_offset(skb) >= packet_buffer_size &&
 	    skb->fclone == SKB_FCLONE_UNAVAILABLE) {
@@ -1945,6 +1946,7 @@ static int octeon3_eth_ndo_start_xmit(struct sk_buff *skb, struct net_device *ne
 			node = gaura >> 10;
 			aura = gaura & 0x3ff;
 			buffers_needed = aura2bufs_needed[node][aura];
+			buf[SKB_AURA_OFFSET] = NULL;
 		}
 	}
 
@@ -1967,11 +1969,11 @@ static int octeon3_eth_ndo_start_xmit(struct sk_buff *skb, struct net_device *ne
 		}
 	}
 
-	/* We have space for 13 segment pointers, If there will be
+	/* We have space for 11 segment pointers, If there will be
 	 * more than that, we must linearize.  The count is: 1 (base
 	 * SKB) + frag_count + nr_frags.
 	 */
-	if (unlikely(skb_shinfo(skb)->nr_frags + frag_count > 12)) {
+	if (unlikely(skb_shinfo(skb)->nr_frags + frag_count > 10)) {
 		if (unlikely(__skb_linearize(skb)))
 			goto skip_xmit;
 		frag_count = 0;
-- 
1.9.1

